source/crud/crud.rst

```diff
@@ -1476,25 +1488,58 @@ Drivers SHOULD report errors however they report other server errors: by raising
 
 Drivers MUST construct a ``WriteConcernError`` from a server reply as follows:
 
-- set ``code`` to ``writeConcernError.code``.
-- set ``message`` to ``writeConcernError.errmsg`` if available.
-- set ``details`` to ``writeConcernError.errInfo`` if available. Drivers MUST NOT parse inside ``errInfo``.
+- Set ``code`` to ``writeConcernError.code``.
+- Set ``message`` to ``writeConcernError.errmsg``.
+- Set ``details`` to ``writeConcernError.errInfo`` if available. Drivers MUST NOT parse inside ``errInfo``.
+
+See the `Read/Write Concern specification </source/read-write-concern/read-write-concern.rst#writeconcernerror-examples>`_
+for examples of how a server represents write concern errors in replies.
+
+WriteError
+~~~~~~~~~~
+
+Write errors for ``insert``, ``update``, and ``delete`` commands are reported as
+objects within a ``writeErrors`` array field in the command response. Drivers
+MUST construct a ``WriteError`` from a server reply as follows (where
+``writeErrors[]`` refers to a particular element in the array):
+
+- Set ``code`` to ``writeErrors[].code``.
+- Set ``message`` to ``writeErrors[].errmsg`` if available.
+- Set ``details`` to ``writeErrors[].errInfo`` if available. Drivers MUST NOT parse inside ``errInfo``.
 
-See the `Read/Write Concern specification </source/read-write-concern/read-write-concern.rst#Errors>`_ for examples of how a server represents write concern errors in replies.
+For single-statement writes (i.e. ``insertOne``, ``updateOne``, ``updateMany``,
+``replaceOne``, ``deleteOne``, and ``deleteMany``), a single write error may be
+reported in the array and ``writeErrors[0].index`` will be zero.
+
+For multi-statement writes (i.e. ``insertMany`` and ``bulkWrite``), potentially
+many write errors may be reported in the array and the ``index`` property will
+be set accordingly. Since the reported ``index`` is specific to each command,
+drivers MUST adjust the index accordingly for ``BulkWriteError.index``.
 
 .. code:: typescript
 
   class WriteError {
 
     /**
-     * An integer value identifying the error.
+     * An integer value identifying the write error. Corresponds to the
+     * "writeErrors[].code" field in the command response.
      *
      * @see https://docs.mongodb.com/manual/reference/method/WriteResult/
      */
     code: Int32;
 
     /**
-     * A description of the error.
+     * A document providing more information about the write error (e.g. details
+     * pertaining to document validation). Corresponds to the
+     * "writeErrors[].errInfo" field in the command response.
+     *
+     * @see https://docs.mongodb.com/manual/reference/method/WriteResult/
+     */
+    details: Document;
```

Comment:
This is the crucial spec change that allows drivers to expose the reason documents fail validation.

That said, there are no tests for this as the unified test format does not support assertions for specific error types beyond [`expectedError.expectResult`](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#expectederror), which is used for asserting the incomplete WriteResult within a BulkWriteException. That was implemented mainly because we already had legacy tests making those assertions.

I'm creating DRIVERS-TBD to track better assertions for WriteError(s) and WriteConcernError, but I don't think we need to complicate an otherwise trivial spec change this late into the release cycle with test runner changes. While something like `expectedError: { hasDetails: true }` could get the job done, I think it'd be hastily designed and likely something we'd come to regret.

----

As far as drivers actually testing this change, I intend to provide some guidance in the description of DRIVERS-820 to suggest that drivers should create a prose test that creates a collection with documentation validation and inserts an invalid document to check that the new error property gets set. My previous investigation in [DRIVERS-1643](https://jira.mongodb.org/browse/DRIVERS-1643) should be helpful for that.

Reply:
Agree with not making more changes to the unified test format this late in the cycle. A single prose test to ensure the property gets populated should be enough until we can make the spec change to the test format.

Reply:
Ah, I didn't realize the CRUD spec had existing prose tests. I'll add something similar to the existing ["errInfo" is propagated](https://github.com/mongodb/specifications/tree/master/source/crud/tests#errinfo-is-propagated) test in a bit.

Reply:
New prose test in https://github.com/mongodb/specifications/pull/1004/commits/24177fac949c0f9ec732a7f587c84668f6eb6c2c